Bluetooth: Controller: llcp: CTE request control procedure fixes

ppryga-nordic opened the pull request:
The PR provides a number of fixes for code related with CTE request control procedure:
- fix to avoid possibility to call HCI_LE_Connection_CTE_Request_Enable with enable parameter set to true when it is already enabled
- use PHY_RX instead of PHY_TX when checking enabled type of PHY to enable CTE request
- add missing CTE request pause when remote PHY update is handled
- add missing CTE request disable implementation
- add missing tests for collisions of CTE request and PHY update
- add missing state "wait for notification buffers" in CTE request state machine
- fix not building `dt/connection_cte_req` tests
- add missing positive behavior tests in `dt/connection_cte_req`

ppryga-nordic committed:
Bluetooth: Controller: df: Add missing CTE request is_enabled set

There were missing set of is_enabled member of CTE request.
That allowed host to run HCI_LE_Connection_CTE_Request_Enable
command with enable parameter set to true sequently.
That violates BT 5.3 Core specification Vol 4 Part B section
7.8.8.85.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: df: refactor CTE REQ set disable

Add a function that is responsible for set CTE request disabled.
It is a refactoring, to have a single place where CTE request
is_enabled and req_interval are zeroed.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: df: Fix wrong PHY type check in CTE req enable

CTE request may not be enabled when actual PHY does not support
transmission of constant tone extension. In case a connection is
asymetric, receive PHY is required to support transmission of CTE.
Receive PHY will be used by a device to get PDU with CTE.

In ll_df_set_conn_cte_req_enable was checking TX PHY type.
It should check RX PHY type.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: llcp: add complete evt handling in idle state

Add handling of complete event in idle state of localy innitiated
control procedures.
The transition will be used by CTE request disable if PHY changes
to CODED while the procedure is waiting in the waiting queue.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: df: Remove not needed if statement

Removes a if statement that was always true.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: df: Add clarifying comment

Add comment to clarify what is a disable_cb callback purpose
and when it is set.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: llcp: Add missing CTE pause in remote PHY change

There were missing pause of CTE request in case remote PHY change
was handled. That may end in violation of BT 5.3 Core Vol 6, Part B
section 5.1.10.1 Packet transmission restrictions.
Without the change there was a possibility to transmitt CTE when
PHY is about to change to PHY CODED.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: Add CTE REQ disable when change PHY to CODED

Accodring to BT 5.3 Core Vol 4, Part E section 7.8.85 the CTE request
procedure has to be disabled as if Host issued the HCI_LE_Connection-
CTE_Request_Enable with enable property set to 0. It means a Controller
should automaticall  disable the procedure.

There were no handling of this part of specification.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: llcp: Add CTE REQ and PHY change collision tests

To avoid violation of BT 5.3 Core Vol 6, Part B section 5.1.10.1
there was added a command pause mechanism that allows to postpone
handling of CTE REQ if there is pending PHY change procedure or
PHY change if there is pending CTE REQ procedure.

The commit adds unit tests for the functionality.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: llcp: Add wait for nrf state in CTE REQ

There were no wait for free notification buffer (pdu object
used to send notifications to Host).
That may cause assertion in lp_comm_ntf function.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: df: Fix not building connection CTE req tests

The tests for direction finding connected mode were not building
correclty. There were missing Kconfig options that enable missing
functionality to be tested.
Tests were not failing because there were no expected positive
behavior tests added.
Added changes allow to validate all implemented test cases.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: df: Add CTE REQ expected positive behavior tests

Add expected positive behavior tests for direction fingin in connected
mode.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

Thalley commented on tests/bluetooth/controller/ctrl_cte_req/src/main.c:
```diff
@@ -616,6 +616,751 @@ void test_cte_req_reject_inv_ll_param_peripheral_remote(void)
+#define HOST_INITIATED 1U
+#define PHY_UPDATE_INSTANT_DELTA 6
+
+#define CHECK_PREF_PHY_STATE(_conn, _tx, _rx)                                                      \
+	do {                                                                                       \
+		zassert_equal(_conn.phy_pref_rx, _rx,                                              \
+			      "Preferred RX PHY mismatch %d (actual) != %d (expected)",            \
+			      _conn.phy_pref_rx, _rx);                                             \
+		zassert_equal(_conn.phy_pref_tx, _tx,                                              \
+			      "Preferred TX PHY mismatch %d (actual) != %d (expected)",            \
+			      _conn.phy_pref_tx, _tx);                                             \
+	} while (0)
+
+#define CHECK_CURRENT_PHY_STATE(_conn, _tx, _flags, _rx)                                           \
+	do {                                                                                       \
+		zassert_equal(_conn.lll.phy_rx, _rx,                                               \
+			      "Current RX PHY mismatch %d (actual) != %d (expected)",              \
+			      _conn.lll.phy_rx, _rx);                                              \
+		zassert_equal(_conn.lll.phy_tx, _tx,                                               \
+			      "Current TX PHY mismatch %d (actual) != %d (expected)",              \
+			      _conn.lll.phy_tx, _tx);                                              \
+		zassert_equal(_conn.lll.phy_rx, _rx,                                               \
+			      "Current Flags mismatch %d (actual) != %d (expected)",               \
+			      _conn.lll.phy_flags, _flags);                                        \
+	} while (0)
```
Do these need to macros instead of functions?
Refer to https://www.kernel.org/doc/html/v4.10/process/coding-style.html#macros-enums-and-rtl

> Generally, inline functions are preferable to macros resembling functions.

And 
> Things to avoid when using macros:
    1. macros that affect control flow:

ppryga replied:
Changed to functions.

Thalley commented on tests/bluetooth/controller/ctrl_cte_req/src/main.c:
```diff
@@ -616,6 +616,751 @@ void test_cte_req_reject_inv_ll_param_peripheral_remote(void)
+
+static bool is_instant_reached(struct ll_conn *conn, uint16_t instant)
+{
+	return ((event_counter(conn) - instant) & 0xFFFF) <= 0x7FFF;
```
`0xFFFF` unnecessary for `uint16_t`. 

What does `0x7FFF` represent?

ppryga replied:
This is maximum positive difference between actual value of connection event counter and instant. It checks if the instant is in the past. The operation is based on 65536 modulo arithmetic. It comes from BT 5.3. Core Vol 6, Part B section 5.5.1.

Thalley commented on tests/bluetooth/controller/ctrl_cte_req/src/main.c:
```diff
@@ -616,6 +616,751 @@ void test_cte_req_reject_inv_ll_param_peripheral_remote(void)
+	struct lll_conn *lll;
+	uint16_t event_counter;
+
+	/* TODO(thoh): Lazy hardcoded */
+	uint16_t lazy = 0;
```
What is this used for?

ppryga replied:
The variable is not used. It was related with peripheral latency but it is not needed in this set of unit tests. Removed.

Thalley commented on tests/bluetooth/controller/ctrl_cte_req/src/main.c:
```diff
@@ -616,6 +616,751 @@ void test_cte_req_reject_inv_ll_param_peripheral_remote(void)
+	ull_conn_default_tx_octets_set(251);
+	ull_conn_default_tx_time_set(2120);
+	ull_dle_init(&conn, PHY_1M);
+	/* Emulate different remote numbers to trigger update of eff */
+	conn.lll.dle.remote.max_tx_octets = PDU_DC_PAYLOAD_SIZE_MIN * 3;
+	conn.lll.dle.remote.max_rx_octets = PDU_DC_PAYLOAD_SIZE_MIN * 3;
```
Why are we multiplying `PDU_DC_PAYLOAD_SIZE_MIN` by 3? 

Do we risk this ever becoming larger than a `PDU_DC_PAYLOAD_SIZE_MAX` if exists?

ppryga replied:
This value is arbitrary. It is used here just to provide some values for PHY update procedure. The tests is not about the correctness of PHY update. Hence the value is just copied from PHY update unit tests.

Thalley commented on tests/bluetooth/controller/ctrl_cte_req/src/main.c:
```diff
@@ -616,6 +616,751 @@ void test_cte_req_reject_inv_ll_param_peripheral_remote(void)
+					   uint8_t ctx_num_at_end)
+{
+	struct pdu_data_llctrl_length_rsp length_ntf = {
+		3 * PDU_DC_PAYLOAD_SIZE_MIN,
+		PDU_DC_MAX_US(3 * PDU_DC_PAYLOAD_SIZE_MIN, phy_req->tx_phys),
+		3 * PDU_DC_PAYLOAD_SIZE_MIN,
+		PDU_DC_MAX_US(3 * PDU_DC_PAYLOAD_SIZE_MIN, phy_req->rx_phys)
```
Here we are reusing the `3 * PDU_DC_PAYLOAD_SIZE_MIN`. 

If they need to be the same value as the above, then we should put this in a macro or a `const` value, and use that across both functions.

In any case, this struct should use a variable or a define for it, as it is using it 4 times.

ppryga replied:
Changed to macro.

Thalley commented on tests/bluetooth/controller/ctrl_cte_req/src/main.c:
```diff
@@ -616,6 +616,751 @@ void test_cte_req_reject_inv_ll_param_peripheral_remote(void)
+				   struct pdu_data_llctrl_phy_req *phy_req, uint8_t events_at_start,
+				   uint8_t ctx_num_at_end)
+{
+	struct pdu_data_llctrl_phy_req rsp = { .rx_phys = PHY_1M | PHY_2M | PHY_CODED,
+					       .tx_phys = PHY_1M | PHY_2M | PHY_CODED };
```
If the PHYs here need to be the same as the PHYs in lines 675-676, consider moving to a common variable or macro.

ppryga replied:
Changed to macro.

Thalley commented on tests/bluetooth/controller/ctrl_cte_req/src/main.c:
```diff
@@ -616,6 +616,751 @@ void test_cte_req_reject_inv_ll_param_peripheral_remote(void)
+				      struct pdu_data_llctrl_phy_req *phy_req,
+				      uint8_t events_at_start, uint8_t ctx_num_at_end)
+{
+	struct pdu_data_llctrl_phy_req rsp = { .rx_phys = PHY_1M | PHY_2M | PHY_CODED,
+					       .tx_phys = PHY_1M | PHY_2M | PHY_CODED };
```
Same comment regarding common PHY values

ppryga replied:
Changed to macro.

Thalley commented on tests/bluetooth/controller/ctrl_cte_req/src/main.c:
```diff
@@ -616,6 +616,743 @@ void test_cte_req_reject_inv_ll_param_peripheral_remote(void)
+	test_phy_update_wait_for_remote_cte_req_complete(BT_HCI_ROLE_PERIPHERAL);
+}
+
+void test_cte_req_wait_for_remote_phy_update_complete_and_disable(uint8_t role)
```
Could these common tests that take a `role` parameter be made `static`?

ppryga replied:
Yes, changed to static.

Thalley reviewed (changes requested):
Appreciate the well documented test cases :)

ppryga-nordic committed:
Bluetooth: Controller: llcp: Fix failing UT for version exchange

Test that verifies correct handling of Version exchanged procedure
called twice on the same connection failed. The reason was wrong
number of expected free procedure context objects. It was expected
that there will be one context not released.
That was wrong because the context is released just after the procedure
completes in idle state. The second and following calls to the procedure
do not start any PDU exchange, so the procedure ends immediately.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Resolved comments from @Thalley.
Added commit that fixes issue in Version exchange procedure unit tests (@kruithofa @thomas-ot have a look at it: a17dffeac34c420f6c134b29564a7565c987a628).

Thalley commented on tests/bluetooth/controller/ctrl_cte_req/src/main.c:
```diff
@@ -616,6 +616,749 @@ void test_cte_req_reject_inv_ll_param_peripheral_remote(void)
+	 *       maximum positive difference between actual value of connection event counter and
+	 *       instant.
+	 */
+	return (uint16_t)(event_counter(conn) - instant) <= (uint16_t)0x7FFF;
```
Is the first cast to `uint16_t` necessary? 

As far as I can tell, `event_counter` returns an `uint16_t` and then you have `uint16_t - uint16_t` which should always return a `uint16_t`, right?

Thalley commented on tests/bluetooth/controller/ctrl_cte_req/src/main.c:
```diff
@@ -616,6 +616,749 @@ void test_cte_req_reject_inv_ll_param_peripheral_remote(void)
+	zassert_equal(conn->lll.phy_tx, phy_tx,
+		      "Current TX PHY mismatch %d (actual) != %d (expected)", conn->lll.phy_tx,
+		      phy_tx);
+	zassert_equal(conn->lll.phy_rx, phy_rx,
```
```suggestion
	zassert_equal(conn->lll.phy_flags, flags,
```

Thalley commented on tests/bluetooth/controller/ctrl_cte_req/src/main.c:
```diff
@@ -616,6 +616,749 @@ void test_cte_req_reject_inv_ll_param_peripheral_remote(void)
+	/* Emulate different remote numbers to trigger update of effective max TX octets and time.
+	 * Numbers are taken arbitrary.
+	 */
+	conn.lll.dle.remote.max_tx_octets = PDU_MAX_TX_RX_OCTETS;
+	conn.lll.dle.remote.max_rx_octets = PDU_MAX_TX_RX_OCTETS;
```
Consider
```suggestion
	conn.lll.dle.remote.max_tx_octets = PDU_MAX_OCTETS;
	conn.lll.dle.remote.max_rx_octets = PDU_MAX_OCTETS;
```

cvinayak reviewed (changes requested):
Minor change request.

ppryga commented on tests/bluetooth/controller/ctrl_cte_req/src/main.c:
```diff
@@ -616,6 +616,749 @@ void test_cte_req_reject_inv_ll_param_peripheral_remote(void)
+	 *       maximum positive difference between actual value of connection event counter and
+	 *       instant.
+	 */
+	return (uint16_t)(event_counter(conn) - instant) <= (uint16_t)0x7FFF;
```
Unfortunately the first cast it also needed. It is caused by integer conversion rules (or integer promotion rules).
In this case we have `event_counter(conn) - instant`  the result of this operation is `int` (at least according to my knowledge). It would have been converted to other type if there is e.g. assignment: `uint16_t value = event_counter(conn) - instant;.

In our case we have: `(event_counter(conn) - instant) <= (uint16_t)0x7FFF` -> `int <= (uint16_t)`.
in this case the `uint16_t` is promoted to `int`, because rank of `int` is greater than rank of `short int` and rank of `short int` it the same as rank of `uint16_t`. At the end of the day, the result of the expression is always `true`.

ppryga replied:
With previous version of the code, one that you suggest to get back, the same rules apply. This is why there is an `& 0xFFFF` in the expression. Similar thing is applied with casts to `uint16_t`.

ppryga commented:
Requested changes done.
Rebased on main branch due to failing CI.
